tsinfer/inference.py

```diff
@@ -1924,6 +1912,156 @@ def get_mutations(self, node_id):
+    if last_edge.left != 0 or last_edge.right != tables.sequence_length:
+        return False  # edge does not span the entire genome
+
+    logger.info("Removing the oldest edge to detach the virtual-root-like ancestor")
+    tables.edges.truncate(tables.edges.num_rows - 1)
+    # move any mutations above the virtual-root-like ancestor to above the grand
+    # MRCA instead (these will be mutations placed by parsimony)
+    mutations_node = tables.mutations.node
+    mutations_node[mutations_node == last_edge.parent] = last_edge.child
+    tables.mutations.node = mutations_node
+    return True
+
+
+def split_genomewide_mrca(tables):
+    # If a single oldest node is a root across the entire genome, split it up into
+    # multiple nodes all with the same node.time, whenever the node children change
+    tables.sort()
+    ts = tables.tree_sequence()
+    # Check that we do have a single root everywhere - there's probably a more
+    # efficient way to do this using the quintuply linked data structure
+    roots = {root for tree in ts.trees() for root in tree.roots}
+    if len(roots) > 1:
+        return False
+
+    logger.info("Splitting the all zeros grand MRCA into separate nodes")
+    genomewide_mrca_id = roots.pop()
+    assert genomewide_mrca_id == tables.edges[tables.edges.num_rows - 1].parent
+    genomewide_mrca = tables.nodes[genomewide_mrca_id]
+    edges_to_remove = tables.edges.parent == genomewide_mrca_id
+    root_breaks = np.unique(
+        np.concatenate(
+            (tables.edges.left[edges_to_remove], tables.edges.right[edges_to_remove])
+        )
+    )
+    assert root_breaks[0] == 0
+
+    # Remove the edges connecting the grand mrca to all its children
+    tables.edges.truncate(tables.edges.num_rows - np.sum(edges_to_remove))
+
+    # We need to move the mutations above the grand_mrca to the new nodes
+    mutation_ids = np.where(tables.mutations.node == genomewide_mrca_id)[0]
+    mutation_positions = tables.sites.position[tables.mutations.site[mutation_ids]]
+    mut_iter = zip(mutation_ids, mutation_positions)
+    trees_iter = ts.trees()
+
+    # Go through the trees, making a new root node whereever we hit a root_break
+    # and recreating the edges to the children each time
+    left = 0
+    mutation_id, mutation_pos = next(mut_iter, (None, ts.sequence_length))
+    tree = next(trees_iter)
+    for right in root_breaks[1:]:
+        while tree.interval.right != right:
+            tree = next(trees_iter)
+        new_root = tables.nodes.append(genomewide_mrca)
+        for c in tree.children(genomewide_mrca_id):
+            tables.edges.add_row(parent=new_root, child=c, left=left, right=right)
+        while mutation_pos < right:
+            tables.mutations[mutation_id] = tables.mutations[mutation_id].replace(
+                node=new_root
+            )
+            mutation_id, mutation_pos = next(mut_iter, (None, ts.sequence_length))
+        left = right
+    return True
```

Comment:
What's the return for?

Reply:
Removed anyway with new code arrangement